### DISCUSSION TITLE
MacOS: Problem with run aiosqlite for any DB

### DISCUSSION THREAD

**Discussion by NeoSPU** [fastapi/fastapi#6640] [Questions] [answered]
### MacOS (macOS Big Sur, version 11.1 (20C69):

**ModuleNotFoundError: No module named 'aiosqlite'**

**Python 3.9 ===== FastAPI ===== SQLite ===== Databases ===== aiosqlite**

### Description
Test project on FastAPI - including from the example: 
[https://fastapi.tiangolo.com/advanced/async-sql-databases/](url)

### Example My Code:
```Python
from typing import List

import databases
import sqlalchemy
from fastapi import FastAPI
from pydantic import BaseModel


DATABASE_URL = "sqlite:///./test.db"

database = databases.Database(DATABASE_URL)
metadata = sqlalchemy.MetaData()


notes = sqlalchemy.Table(
    "notes",
    metadata,
    sqlalchemy.Column("id", sqlalchemy.Integer, primary_key=True),
    sqlalchemy.Column("text", sqlalchemy.String),
    sqlalchemy.Column("completed", sqlalchemy.Boolean),
)


# engine = sqlalchemy.create_engine(
#     DATABASE_URL, connect_args={"check_same_thread": False}
# )
# metadata.create_all(engine)

app = FastAPI()


@app.get("/")
async def root():
    return {"message": "Hello World"}
```

#### Start: 
Activate virtual environment:
```Shell
$ pip3 -m venv venv
$ source ./venv/bin/activate
```

Set dependencies:
```Shell
$ pip3 install aiosqlite, databases, fastapi
```

Look the installed dependencies:
```Shell
$ pip3 list
Package           Version
----------------- --------
aiosqlite         0.17.0
databases         0.4.3
fastapi           0.63.0
pip               21.1.1
pydantic          1.8.1
setuptools        56.0.0
SQLAlchemy        1.3.24
starlette         0.13.6
typing-extensions 3.10.0.0
```

Start server:
`$ uvicorn main:app --reload`

The server won't start On Mac OS (any editor: vim, VS Code, "Pycharm", etc.)
### ERROR shell:
```SHELL
(env) ➜  testDB2 uvicorn main:app --reload
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [2967] using statreload
Process SpawnProcess-1:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/uvicorn/subprocess.py", line 62, in subprocess_started
    target(sockets=sockets)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/uvicorn/main.py", line 390, in run
    loop.run_until_complete(self.serve(sockets=sockets))
  File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/uvicorn/main.py", line 397, in serve
    config.load()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/uvicorn/config.py", line 278, in load
    self.loaded_app = import_from_string(self.app)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/uvicorn/importer.py", line 23, in import_from_string
    raise exc from None
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/uvicorn/importer.py", line 20, in import_from_string
    module = importlib.import_module(module_str)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 855, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/Users/alexneo/Documents/GIT/FastAPI/testDB2/./main.py", line 11, in <module>
    database = databases.Database(DATABASE_URL)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/databases/core.py", line 66, in __init__
    backend_cls = import_from_string(backend_str)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/databases/importer.py", line 21, in import_from_string
    raise exc from None
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/databases/importer.py", line 18, in import_from_string
    module = importlib.import_module(module_str)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 855, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/databases/backends/sqlite.py", line 5, in <module>
    import aiosqlite
ModuleNotFoundError: No module named 'aiosqlite'
```
**All of the dependencies are exist and I've can click them and moved to these dependencies file!!!**
**And 'aiosqlite' too!**

**on Linux, everything works OK**

but on a MacOS I can not run any of the databases (example SQLITE)

The error is as follows: does not see "aiosqlite"! All of the requirements have been seated correct (Python 3.9.5, venv, pip, aiosqlite, sqlalchemy, databases etc. - all of this have been updated)

### This problem at StackOverflow:
 [https://stackoverflow.com/questions/67406165/macos-problem-with-run-aiosqlite-for-any-db](url)

### This solution hasn't worked to my fail:
[https://question-it.com/questions/467425/kak-ja-mogu-zapustit-server-fast-api-ispolzuja-pycharm](url)

### Screenshots:
* Error [https://i.stack.imgur.com/r50bg.jpg](url)
* Dependencies [https://i.stack.imgur.com/OeBD9.jpg](url)
* Pip List [https://i.stack.imgur.com/KbFRB.jpg](url)

**Comment by akx** [comment]
Based on the line

> /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/databases/backends/sqlite.py

you have at least the `databases` package installed outside your virtualenv. You should clean up the non-virtualenv site-packages – it'll be darned difficult to debug any issue if a library is mixing and matching modules from various paths.

**Comment by NeoSPU** [comment]
> Based on the line
> 
> > /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/databases/backends/sqlite.py
> 
> you have at least the `databases` package installed outside your virtualenv. You should clean up the non-virtualenv site-packages – it'll be darned difficult to debug any issue if a library is mixing and matching modules from various paths.

Sorry! But, If We want to find a solution - We must try to research this question - without prejudiced opinion!

**Try doing this with me step-by-step:**

### Step 1
We haven’t any virtual environments and installed packages:
![image](https://user-images.githubusercontent.com/29407496/118369886-e1f79180-b5ad-11eb-8ec5-b8dd5227cf01.png)

### Step 2
We activate Virtual Environment:
``` Shell
$ python3 -m venv venv
$ source ./venv/bin/activate
```
![image](https://user-images.githubusercontent.com/29407496/118369941-14a18a00-b5ae-11eb-906b-f6824aa6a158.png)

### Step 3
We’ve installed "fastapi" package:
``` Shell
$ pip3 install fastapi
```
![image](https://user-images.githubusercontent.com/29407496/118369975-40247480-b5ae-11eb-92e5-57fd5d1df720.png)

### Step 4
We haven’t installed **“dtabases”** package or **“aiosqlite”!**
If We'll try to run our server - We’ve to sow next error message:

![image](https://user-images.githubusercontent.com/29407496/118370018-6b0ec880-b5ae-11eb-96b5-37d68092d4d1.png)

### Step 5
We'll try to install “aiosqlite”at our virtual environment:
```Shell
$ pip3 install “aoisqlite”
```

![image](https://user-images.githubusercontent.com/29407496/118370051-8f6aa500-b5ae-11eb-8f72-e32e50dad1cd.png)

### Step 6
Looking at our virtual environment library:

![image](https://user-images.githubusercontent.com/29407496/118370064-9d202a80-b5ae-11eb-90c9-146c2fdfa24d.png)

![image](https://user-images.githubusercontent.com/29407496/118370068-a3160b80-b5ae-11eb-9715-00f0885541bb.png)

**We've got an installed "aiosqlite" package!**

And We've didn't installed "databases" package!

### Step 7
We’ll try to run our server 
```Shell
$ uvicorn main:app —reload
```
![image](https://user-images.githubusercontent.com/29407496/118370089-c345ca80-b5ae-11eb-92f5-b7f4dafdb604.png)

**And We’ve got an error by “aiosqlite” again!**
Oops!


P.S. I’m apologies, but What do You mean, Dear Aarni Koskela?

Regards,
NeoSPU!

**Comment by akx** [comment]
Dear @NeoSPU:

`databases` isn't part of the standard library, yet you don't install it, and yet you are able to import it and you can see in the tracebacks that its code is being run.) That means that it is somehow available as part of your system site-packages, as I mentioned before. (The same stands for `sqlalchemy` - your code seems to use it, yet you don't install it.)

Please run `python3 -m pip list` _outside_ any virtualenv and let us know the output. On a "venv-clean" system it should have little more than `pip`, `wheel`, `setuptools`.

**Comment by akx** [comment, marked as answer]
To wit: I followed the steps above, with adaptations to my system.

You can see from `/Users/akx/envs/fastapi-3224/lib/python3.9/site-packages/databases/backends/sqlite.py` that the `databases` package is indeed installed _in the venv's_ site-packages, i.e. inside the venv, as things should be.

After installing all of the dependencies (within the venv), your script does not raise an exception.

```
created virtual environment CPython3.9.4.final.0-64 in 391ms
  creator CPython3Posix(dest=/Users/akx/envs/fastapi-3224, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/Users/akx/Library/Application Support/virtualenv)
    added seed packages: pip==20.3.1, setuptools==51.0.0, wheel==0.36.1
  activators BashActivator,CShellActivator,FishActivator,PowerShellActivator,PythonActivator,XonshActivator
Activating /Users/akx/envs/fastapi-3224
(fastapi-3224) ~/b/fastapi-3224 $ cat > code.py
# (elided code from original post)
(fastapi-3224) ~/b/fastapi-3224 $ python code.py
Traceback (most recent call last):
  File "code.py", line 3, in <module>
    import databases
ModuleNotFoundError: No module named 'databases'
(fastapi-3224) ~/b/fastapi-3224 $ pip install databases
Successfully installed databases-0.4.3 sqlalchemy-1.3.24
(fastapi-3224) ~/b/fastapi-3224 $ python code.py
Traceback (most recent call last):
  File "code.py", line 5, in <module>
    from fastapi import FastAPI
ModuleNotFoundError: No module named 'fastapi'
(fastapi-3224) ~/b/fastapi-3224 $ pip install fastapi
Successfully installed fastapi-0.65.1 pydantic-1.8.2 starlette-0.14.2 typing-extensions-3.10.0.0
(fastapi-3224) ~/b/fastapi-3224 $ python code.py
Traceback (most recent call last):
  # ...
  File "/Users/akx/envs/fastapi-3224/lib/python3.9/site-packages/databases/backends/sqlite.py", line 5, in <module>
    import aiosqlite
ModuleNotFoundError: No module named 'aiosqlite'
(fastapi-3224) ~/b/fastapi-3224 $ pip install aiosqlite
Successfully installed aiosqlite-0.17.0
(fastapi-3224) ~/b/fastapi-3224 $ python code.py
(fastapi-3224) ~/b/fastapi-3224 $
```

**Comment by NeoSPU** [comment]
### Yes! 
### Awesome! 
## Thanks, Aarni (Aarni Koskela)! 

All of my questions by issue #3224 have been solved! 

The solution is next: 

**_all dependencies by the database like the "aiosqlite" package must be installed outside the virtual environment!_**

**Comment by NeoSPU** [comment]
Thank You, Very Much, Aarni Koskela!

Regards,
Alex!

**Comment by akx** [comment]
@NeoSPU 

> all dependencies by the database like the "aiosqlite" package must be installed outside the virtual environment!

No, quite the contrary: you should strive to install all dependencies for a project _inside_ the virtual environment. This is why I suggested to clean the outside-virtual-environment... environment.

**Comment by NeoSPU** [comment]
@akx Aarni Koskela

Yes!
### You are right!

### I've fixed my solution!

The solution is next:

- cleaning dependencies, that have been installed outside the virtual environment:
```Shell
$ ~ pip freeze > to_delete.txt
$ ~ pip uninstall -y -r to_delete.txt
$ ~ /usr/local/opt/python@3.9/bin/python3.9 -m pip install --upgrade pip
$ ~ pip list
Package    Version
---------- -------
pip        21.1.1
setuptools 54.2.0
wheel      0.36.2
```
- create a virtual environment:
```Shell
$ ~ python3 -m venv venv
$ ~ source ./venv/bin/activate
- install dependencies from the requirements.txt file into your virtual environment of the project.
```
**requirements.txt:
```
aiosqlite==0.17.0
click==7.1.2
databases==0.4.3
fastapi==0.65.1
h11==0.12.0
pydantic==1.8.2
SQLAlchemy==1.3.24
starlette==0.14.2
typing-extensions==3.10.0.0
uvicorn==0.13.4
```
```Shell
$ ~ pip3 install -r requirements.txt
$ ~ uvicorn main:app --reload
```

### Thank you!

**Comment by tiangolo** [comment]
Thanks for the help here @akx ! 👏  🙇 

Thanks for reporting back and closing the issue @NeoSPU  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
